KEP-3766: Move ReferenceGrant to sig-auth API Group (#3767)
kubernetes/enhancements: keps/sig-network/3766-referencegrant/README.md

}

// ReferenceGrantSpec identifies a cross namespace relationship that is trusted
// for Gateway API.
type ReferenceGrantSpec struct {
	// From describes the trusted namespaces and kinds that can reference the
	// resources described in "To". Each entry in this list MUST be considered
	// to be an additional place that references can be valid from, or to put
	// this another way, entries MUST be combined using OR.
	//
	// +kubebuilder:validation:MinItems=1
	// +kubebuilder:validation:MaxItems=16
	From []ReferenceGrantFrom `json:"from"`

	// To describes the resources that may be referenced by the resources
	// described in "From". Each entry in this list MUST be considered to be an
	// additional place that references can be valid to, or to put this another
	// way, entries MUST be combined using OR.
	//
	// +kubebuilder:validation:MinItems=1
	// +kubebuilder:validation:MaxItems=16
	To []ReferenceGrantTo `json:"to"`
}

// ReferenceGrantFrom describes trusted namespaces and kinds.
type ReferenceGrantFrom struct {
	// Group is the group of the referent.
	// When empty, the Kubernetes core API group is inferred.
	Group Group `json:"group"`

> 	// Kind is the kind of the referent.

Reviewer: where did we settle in the Kind vs Resource recommendation?
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#kind-vs-resource

Reviewer 2: We considered this quite a bit, but erred on the side of being able to use `Kind`, so that people could take the value from the metadata field in the first couple of lines YAML of the object they're referring to, and put it straight into these fields, rather than needing to do a kind-resource lookup for every user's interaction with the API.

As in the Ingress API, the target types must be clearly defined by each implementation, so there should be no opportunity for ambiguity. On top of that, this is only _half_ of a handshake, so the From referent must also both support referencing the To referent, and must reference the one in the To field.

With all of that in mind, we felt that the benefits in terms of specificity outweighed the costs in terms of user pain to use the more correct `Resource`.

Reviewer 2: I added a "Design notes" section that includes this info, resolving.

Reviewer 3: See above comments - I think it needs to be resource

Reviewer 2: Yes, I will update this section accordingly.

- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

[ReferenceGrant](https://gateway-api.sigs.k8s.io/api-types/referencegrant/) was
developed by the [Gateway API subproject](https://gateway-api.sigs.k8s.io/) to
enable certain object references to cross namespaces. More recently, it has also
been [used by
sig-storage](https://kubernetes.io/blog/2023/01/02/cross-namespace-data-sources-alpha/)
to enable cross-namespace data sources.

This KEP proposes moving ReferenceGrant from its current
`gateway.networking.k8s.io` API group to a new `grants.authorization.k8s.io` API
group.

## Motivation

> Now that it's clear that ReferenceGrant is useful beyond just Gateway API, it
> would be good to formalize this model in a more neutral home, ideally sig-auth.
> At this point, each project that wants to enable cross-namespace references has
> to choose between introducing a dependency on Gateway API and creating a new
> resource that would largely duplicate ReferenceGrant. Both options would lead to
> confusion for Kubernetes users.

Reviewer 4: Small nit: I think it might improve readability to reverse these sentences for future readers. I also suggested some changes here too for your considerations:

```suggestion
Any project that wants to enable cross-namespace references currently has to choose
between introducing a dependency on Gateway API's ReferenceGrant or creating a
new API that would be partially redundant (leading to confusion for users).

Recent interest between SIGs has made it clear that ReferenceGrant is wanted for use
cases other than Gateway API. We would like to move ReferenceGrant to a neutral home
(ideally, under sig-auth) in order to make it the canonical API for managing references
across namespaces.
```

Reviewer 2: Thanks, I agree, updating.

### Goals

* Move ReferenceGrant to sig-auth API Group
* Clearly define how ReferenceGrant should be used, including both current use
  cases and guidance for future use cases
* Implement a library to ensure that ReferenceGrant is implemented consistently
  by all controllers

### Non-Goals

* Add, remove, or change fields in the ReferenceGrant API
* Develop an authorizer that will automatically implement ReferenceGrant for

>   all use cases

Author: I think that would be incompatible with some of the requests we're seeing elsewhere in this KEP, such as "all namespaces" - I think label selectors would be especially problematic for this kind of approach.

>   all use cases

Reviewer 5: RBAC does support `*` but not selectors.  This might still be worth discussing though because it could drastically reduce code complexity outside of core.

## Proposal

Move the existing ReferenceGrant resource into a new

> `grants.authorization.k8s.io` API group, defined within the Kubernetes code base

Reviewer 6: do we anticipate any other objects in this API group? would keeping this in the `authorization.k8s.io` API group make sense?

Author: Not tied to any API group, just added this as a starting point, will adjust to `authorization.k8s.io`.

> `grants.authorization.k8s.io` API group, defined within the Kubernetes code base

Author: Looks like comments elsewhere settled on `authorization.k8s.io` which we've updated to.

as part of the 1.27 release.

We will take this opportunity to clarify underspecified parts of the API, but
will not add, change, or remove any fields as part of this transition.
This resource will start with v1beta1 as the API version, matching the API version

> it already has within Gateway API.

Reviewer 2: We didn't mention this inline, but, since Gateway API is a CRD using the `kubernetes.io` API group, ReferenceGrant has already had a Kubernetes API review pass to move from Alpha to Beta, that's why we're proposing it go straight to beta.

I can see that maybe sig-auth review is different, so maybe it does make sense to move this back to alpha. I'll update this part of the text to clarify that it's already gone through a general Kubernetes API review though.

> it already has within Gateway API.

Reviewer 5: I think moving to a new SIG and becoming a general API dramatically changes the context of the API review itself.  Maybe I am being too cautious...

> it already has within Gateway API.

Reviewer 7: I definitely have some feedback on this so I think Alpha is warranted. This type of thing can cause significant auth bugs.

> it already has within Gateway API.

Reviewer 3: after reading it, I doubt this API survives without any changes :)

This could lead to confusion for users. We'll need to rely heavily on
documentation for this feature, tracking all uses of official Kubernetes APIs
that support ReferenceGrant in a central place.

#### Potential for Variations Among Implementations
Because this relies on each individual controller to implement the logic,
it is possible that implementations may become inconsistent. To avoid that,
we'll provide a standard library for implementing ReferenceGrant. We'll
also strongly recommend that every API that relies on ReferenceGrant 
includes robust conformance tests covering this functionality. Existing
Gateway API conformance tests can serve as a model for this.

#### Cross-Namespace References may Weaken Namespace Boundaries
Although we believe that the handshake required for cross-namespace references
with ReferenceGrant ensures these references will be safe, it does potentially
weaken existing namespace boundaries. We believe ReferenceGrant will have a net
benefit on the ecosystem as it will allow workloads, secrets, and configuration
to be deployed in separate namespaces that more clearly match up with desired
authorization.

## Design Details

### General Notes

> #### `ReferenceGrant` is half of a handshake

Reviewer 6: hoisting from the [sig-auth discussion](https://docs.google.com/document/d/1woLGRoONE3EBVx-wTb4pvp4CI7tmLZ6lS26VTbosLKM/edit#bookmark=id.9q15hi5lip46):

does anything ensure the user creating the ReferenceGrant has permissions (read? write?) on the object they are granting access to? Translating the existing ReferenceGrant into an authz check means translating from Kind to Resource

since this is extending "half of a handshake", it seems important to ensure the actor extending the handshake actually has permission to do so

Reviewer 5: This reminds me of a question I forgot: who can create / mutate these grant objects?  Is it part of the `edit` or `admin` cluster roles?

+1 on having a coverage check similar to what RBAC does today.

Author: Yes, ReferenceGrant is designed to be revocable, though I think this particular kind of check that the creator/updater of ReferenceGrant indeed has read or write access to the resources it grants access to would only be reasonable at the time the ReferenceGrant was created/updated. If that user had their read/write access revoked, I don't think it is reasonable to revoke the ReferenceGrant previously created by them.

Reviewer 2: I agree that the ideal solution here is that _something_ verifies that the identity creating the ReferenceGrant has access to all the resources it is granting access to.

However, both this and other changes of this kind sound like either a validating webhook or apiserver-native validation. It doesn't seem like it would be a good thing for controllers reconciling these resources to need to go and look up who created them. This is an area I don't know much about though.

I'll leave this as is for now.

Author: Added a new section to that states this requirement, for now it just requires read access but can follow up with more specifics on that section.

object).

#### `Kind` vs `Resource`

When creating a metaresource (that is, a resource that targets other resources)
like ReferenceGrant, it's important to consider if the metaresource uses the more
common `Kind` or the more correct `Resource`.

When designing the Gateway API in general, we considered this quite a bit (and
@robscott even ended up making https://github.com/kubernetes/community/pull/5973
to clarify the API conventions), but in the end decided to use Kind, to improve
the user experience. That is, it's easier users to take the value from the `kind`
field at the top of the YAML they are already using, and put it straight into
these fields, rather than needing to do a kind-resource lookup for every user's
interaction with the API.

Secondly, as in the Ingress API, the target types must be clearly defined by
each implementation, so there should be no opportunity for ambiguity.

Lastly, this is only half of a handshake, so the From referent must also both
support referencing the To referent, and must reference the one in the To field.

With all of that in mind, we felt that the benefits in terms of specificity

> outweighed the costs in terms of user pain to use the more correct `Resource`.

Reviewer 6: Using `kind` instead of `resource` means that any authorization verification that the user creating/modifying the ReferenceGrant actually has permission to access the object they are granting access to has to do a REST mapping lookup, which could fail. Examples at:
* https://github.com/kubernetes/kubernetes/blob/release-1.26/pkg/registry/admissionregistration/validatingadmissionpolicy/authz.go#L72-L83
* https://github.com/kubernetes/kubernetes/blob/release-1.26/plugin/pkg/admission/gc/gc_admission.go#L206-L215

That should at least be mentioned as a downside of using `kind`, and discuss how those error cases should be handled

Reviewer 3: Yeah, I think `Kind` is insufficient here.  It needs to be `Resource`.

Within a very limited domain, Kind is OK - the implementing controller knows that Group-Kind -> Resource is bounded and enumerable (and usually in a group under the same control).  E.g. a Foo Gateway impl allows Kind "FooRoute".  

In the case of RefGrant, a Kind would seem to mean _any_ resource which uses that Kind (I hate that we have this distinction, TBH, but...), and I doubt that is the sort of ambiguity we want in an authz API.

Reviewer 2: Okay, I guess I've lost this one. 😄 

The one advantage I can see of using Resource rather than Kind is that resources are plural, so it will end up more clear that you're granting access from more than one thing.

So, instead of allowing access to Gateways using this one (as we do now in Gateway API):
```yaml
    from:
      # Note that in Gateway API, Group is currently defaulted
      # to this, which means you to explicitly set the group to
      # the empty string for Core resources. We should definitely
      # change this.
      - group: "gateway.networking.kubernetes.io"
        kind: Gateway
        namespace: foo
```

It will be this one:
```yaml
    from:
      # Assuming that we leave the default for Group to the empty
      # string, so that Core objects don't need additional config.
      - group: "gateway.networking.kubernetes.io" 
        resource: gateways
        namespace: foo
```

Which implies that it's _all_ Gateways in the namespace better.

It does suck a bit that the docstring for the field will need to remind people about kind/resource distinctions, but I agree that we need the extra caution here. Sigh.

Reviewer 2: I'll do this update tomorrow my time (finishing up for the day soon).

> ### API Spec

Reviewer 2: That's very fair. I'll make this update.

```golang
// ReferenceGrant identifies kinds of resources in other namespaces that are
// trusted to reference the specified kinds of resources in the same namespace
// as the policy.
//
// Each ReferenceGrant can be used to represent a unique trust relationship.
// Additional ReferenceGrants can be used to add to the set of trusted
// sources of inbound references for the namespace they are defined within.
//
// ReferenceGrant is a form of runtime verification allowing users to assert
// which cross-namespace object references are permitted. Implementations that
// support ReferenceGrant MUST NOT permit cross-namespace references which have
// no grant, and MUST respond to the removal of a grant by revoking the access

> // that the grant allowed.

Reviewer 6: Can this clarify whether that means existing objects that made use of the cross-namespace relationship should be deactivated/torn down, or that they should coast on the last-granted content? For example, if load balancer A in namespace A uses a TLS secret from namespace B authorized via a reference grant, should removing that grant deactivate the load balancer or prevent it from getting updated content from the secret in namespace B?

Reviewer 3: It seems like this will probably be contextual - I think a TLS cert should be revoked ASAP, but the HTTP route itself would not be torn down.

Reviewer 2: In the existing Gateway API implementation, it is contextual - certs being revoked means the Listener becomes invalid and so is removed, but a cross-namespace backend's ReferenceGrant being revoked means _that backend_ is invalid, which only definitely means _that backend_ is removed. But if it's the only backend, then the removal can cascade outwards.

To answer @liggitt's question explicitly: currently in Gateway API, if user A uses a Gateway in namespace A to reference a Secret in namespace B, and the ReferenceGrant is removed, then the Listener is no longer valid, and must be removed. (Having invalid TLS config is a fatal config error).

I'll add something to say both that:
- the removal must be done as soon as possible (eventual consistency permitting)
- the results of the removal _may_ depend on the exact use the reference is being put to, but _should_ err on the side of removing access.

We talked a lot about where to err here for Gateway API, on keeping things up, or having things be more "secure" (for some value of secure), and came down on the side of more secure. As commented elsewhere here, ReferenceGrant is an access control object, removing it is likely to be a destructive operation. (This is another reason why @robscott and I are both in favor of _some_ form of status communication, whether or not it's a `status` subresouce).

Author: I think this is well covered with the latest updates to the KEP, let me know if I missed anything.

> // that the grant allowed.

Author: Yep, added some clarification here, PTAL.

//
// Support: Core
type ReferenceGrant struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	// Spec defines the desired state of ReferenceGrant.

> 	Spec ReferenceGrantSpec `json:"spec,omitempty"`

Reviewer 3: Do we think we will really have status?

Author: I would like to be able to have a status that indicates which references this resource has enabled and for which implementations. Maybe a simple structure like this:

```yaml
status:
  referencesAllowed:
  - controllerName: gateway.example.com
    numReferences: 1
```

In general we want users to be able to understand if a ReferenceGrant is actually being used or if it can be safely deleted. It would be even better if we could describe the references it enabled, but that would almost certainly not be scalable.

Reviewer 3: This seems like an unbounded list problem.  I guess you could truncate.

You'll need participating controllers to a) have authz for refgrant/staus (probably OK, but a bit scary); and b) periodically refresh their name and count in this list; and c) clean up when the ref is not used any more.  It sounds fragile and racy.

I ACK the desire for visibility (applaud it, actually) but I am not sure this is the right design.

Perhaps what you want is a RefGranted resource which indicates a single reference-grant-use, or maybe even just a periodic event saying "this refgrant is in use by gateway foo in namespace bar".  That's easy to periodically throw, doesn't have an unbounded list, doesn't have RBAC issues or cleanup issues.

I don't know if an accurate tally really matters here, as much as a "someone is using this" ? 

If we're going to bake spec/status into the API, we should believe it is needed, so maybe a bit of legwork on this aspect?

Author: > Perhaps what you want is a RefGranted resource which indicates a single reference-grant-use or maybe even just a periodic event saying "this refgrant is in use by gateway foo in namespace bar".

I'm a bit concerned with how noisy events or additional resources would get here. In the case of events, it seems like every controller would need to publish an event for each ReferenceGrant it was using at least 1x an hour. In the case of resources, the number of ReferenceGrantUse resources could become massive. 

If we're using status, I think the unbounded list is slightly less of an issue when the list is just the number of distinct implementations of the API. That seems highly likely to be <100 for a long time, but I could be wrong. We could also take the struct I'm suggesting putting in status and making a resource with just that for ReferenceGrantUse...

>I don't know if an accurate tally really matters here, as much as a "someone is using this" ?

I mostly agree, but a tally could help someone understand the order of magnitude + if they were successfully reducing the number of uses in preparation for a delete.

Reviewer 8: > ```yaml
> status:
>   referencesAllowed:
>   - controllerName: gateway.example.com
>     numReferences: 1
> ```

- If this kind of behavior is important, maybe what a finalizer could do the right job? (If you don't want a controller to be able to set that finalizer, you can choose not to let it).
- I'm thinking that if we want a controller to be able to set `status.referencesAllowed`, we could have a matching, _optional_ declaration within `.spec` about which controller IDs this reference grant applies to.

We don't need `.status` right now, just room to add it later.

Reviewer 3: > I'm a bit concerned with how noisy events or additional resources would get here. In the case of events, it seems like every controller would need to publish an event for each ReferenceGrant it was using at least 1x an hour. In the case of resources, the number of ReferenceGrantUse resources could become massive.

ehhhhh.  We don't have it today, and it took 8 years for us to justify it, so will it REALLY grow that large that quickly?  Events are assumed to be somewhat high N already.  That said, let's define what it is that you're tying to indicate.  Is it "this refgrant was used" or "this refgrant was used approximately N times" or "this refgrant was used EXACTLY N times" or this refgrant is used by Foo X in namespace Y" ?

Those all have different requirements :)

> If we're using status, I think the unbounded list is slightly less of an issue when the list is just the number of distinct implementations of the API

Then a RefGrantUse resource is the same - if it is one per subject per controller, it's the same cardinality, right?

Author: I think I have the following goals, in descending order of importance:

1. Communicate that the ReferenceGrant is actively being used
2. Communicate which controllers are using this ReferenceGrant
3. Communicate how many times it's been used with sufficient granularity that I can see the effects of my changes (if I remove this reference, am I removing a dependency on this ReferenceGrant?)

Both 1 and 2 could be accomplished with a finalizer, and each individual implementation could choose to expose finalizers as an option since I realize that many don't like to deal with finalizers. I'm a bit sad to give up on 3, but if we _really_ want something like that we could introduce a secondary resource like you're describing.

Reviewer 3: If we think that the cardinality of controllers is low, we can put it in status.  The downsides are:
1) Could frequently require retries because of optimisitic concurrency failures (I'm trying to increment my count, but so is everyone else)
2) If we're wrong about cardinality, there's not an easy way out
3) Lots of writes to a resource that will be watched fairly often (every controller which needs refs will watch all refgrants)
4) We need `.status`.

If we instead put that into a `ReferenceGrantUse` resource (just a tuple of controller-name and count), then we only have optimisitic concurrency problems with ourselves, we have ~infinite cardinality, nobody will be watching them, and RefGrant doesn't need .status.

Downsides:
1) It's another new resource
2) It's a new pattern, untested in other places.

Author: Moved to "open questions" since I think we're going to need more discussion here.

	// Note that `Status` sub-resource has been excluded at the
	// moment as it was difficult to work out the design.
	// A `Status` sub-resource may be added in the future.
}

// +kubebuilder:object:root=true
// ReferenceGrantList contains a list of ReferenceGrant.
type ReferenceGrantList struct {
	metav1.TypeMeta `json:",inline"`
	metav1.ListMeta `json:"metadata,omitempty"`
	Items           []ReferenceGrant `json:"items"`
}

// ReferenceGrantSpec identifies a cross namespace relationship that is trusted
// for Gateway API.
type ReferenceGrantSpec struct {
	// From describes the trusted namespaces and kinds that can reference the
	// resources described in "To". Each entry in this list MUST be considered
	// to be an additional place that references can be valid from, or to put
	// this another way, entries MUST be combined using OR.
	//
	// +kubebuilder:validation:MinItems=1
	// +kubebuilder:validation:MaxItems=16

> 	From []ReferenceGrantFrom `json:"from"`

Author: If it's ok to leave them off of the KEP, I'll consider this resolved, but feel free to reopen if we should include them here as well.

> 	// To describes the resources that may be referenced by the resources

Reviewer 3: "describes the resources in this namespace" ?

Is "to" the right term?  We had a discussion a while back about the target of policies and, IIRC, the general trend was "subject" ?  https://groups.google.com/g/kubernetes-api-reviewers/c/ldmrXXQC4G4

Author: I personally think "From" and "To" provide a very clear distinction. I can't think of a great word to pair with "Subject" in the place of "From", but open to ideas.

Reviewer 2: I agree that `subject` is more correct, but the antonym there is probably `object`, and having a ReferenceGrant with `subject` and `object` feels like it would get into... grammatical correctness discussions far more than we would like.

I agree that the language here should be clearer though.

Reviewer 2: I've updated this, PTAL.

Reviewer 3: I would not argue for subject/object - that's confusing, and I love to take analogies too far.

NetPol uses To/From but only for actual communications, and that still has been critiqued as perhaps "too cute".

Subject/From isn't too bad, but not as symmetric.  Subject/Referrer is correct but decidedly uncute.  Subject/Origin ?

I hold opinions from an API review POV, but I'd like sig-auth to own the decision :)

Author: Moved to "open questions" since I think we're going to need more discussion here.

	// described in "From". Each entry in this list MUST be considered to be an
	// additional place that references can be valid to, or to put this another
	// way, entries MUST be combined using OR.
	//
	// +kubebuilder:validation:MinItems=1

> 	// +kubebuilder:validation:MaxItems=16

Reviewer 2: In general, we've tried to err on the side of having _some_ limit for all lists in Gateway API objects, to ensure that no-one runs up against etcd object size limits. For the From and To fields here, we're trying to encourage people to have more, smaller ReferenceGrant objects, so that we don't end up with single objects granting access to many things.

tbh I don't remember why we picked 16 over some other number though.

> 	// +kubebuilder:validation:MaxItems=16

Reviewer 3: you;ve got to start somewhere :)

> 	// +kubebuilder:validation:MaxItems=16

Reviewer 3: Every list field should have a defined limit.

Every string field should have a max size.

	To []ReferenceGrantTo `json:"to"`
}

// ReferenceGrantFrom describes trusted namespaces and kinds.
type ReferenceGrantFrom struct {
	// Group is the group of the referent.
	// When empty, the Kubernetes core API group is inferred.
	Group Group `json:"group"`

	// Kind is the kind of the referent.

> 	Kind string `json:"kind"`

Reviewer 6: hoisting this question from the [sig-auth discussion](https://docs.google.com/document/d/1woLGRoONE3EBVx-wTb4pvp4CI7tmLZ6lS26VTbosLKM/edit#bookmark=id.9q15hi5lip46):

If the same kind can be referenced in two different ways from another API, is group/kind granularity sufficient for the `from` allowlist? As a made up example, if Volumes can reference VolumeSnapshots as a source (readonly) or a destination (write), could I allow my snapshots to be used as a source but not a write destination?

Reviewer 3: Most of the comments in this PR are about this topic :)

Reviewer 2: I agree that it's conceivable that we might want to add extra dimensions to the `from` stanza to further restrict the access in the future, but to me, that sounds like a great place where we could add an optional field in a compatible way later, if we see a demand for it.

I think that the design of any additional dimensions would be pretty tricky too - we'd almost need an annotation-style map[string]string to be able to describe arbitrary access subdivisions. Which give me the yucks, personallly.

Reviewer 6: Taking the volume / volumesnapshot thought experiment further and thinking about safe evolution of references, is it a good idea to leave an indication of the purpose for the reference optional?

Imagine:
1. volumes add the ability to provision from snapshots (read only), so someone creates a ReferenceGrant object to allow referencing all the volumesnapshots in their `templates` namespace with something like `from: [{resource:volumes, group:storage.k8s.io, namespace:*}]` and `to: {resource:volumesnapshots, group:snapshot.storage.k8s.io}`
2. at a later point in time, volumes add the ability to back up to snapshots, qualified by `for: backup` or something in the `from` entries

Would the existing ReferenceGrant that did not scope the purpose for the grant (but assumed it was limited to read-only use, since that was the only use defined at the time of creation) automatically expand to allow any volume in any namespace to reference the snapshots as a write destination for a backup?

Author: I think the most straightforward approach would be for this to only grant read only referential access initially. In the future we could add a new optional field that allowed users to grant write access if such a need ever came up.

On the other hand, one of the current uses, forwarding traffic to a Service in another namespace does not clearly fall in the "read" or "write" category to me. I think I could be OK with a `for` field that allowed for a variety of predefined and/or domain-prefixed values. Since this is already within a list, I'd prefer to keep this as an individual string so we don't end up with a list within a list.

Author: Introducing a `verbs` or `for` field could actually enable us to relatively safely allow `Any` or `*` as a value for resources or groups in `From`. For the sake of consistency and flexibility, `verbs` seems like it could be a good fit. Maybe we could define some well known/standard values as part of this API but also allow domain-prefixed verbs for other uses of the API?

Reviewer 3: Lot's of comments that are now "outdated" which need followup - Github makes this hard.

I agree that read-only is simpler (and I'd include Service forwarding in that), but I *do* think we should ask the larger question of where we want this to evolve.   At one end is a simple resource which only models read-only access.  At the other end is a robust mechanism which governs cross-ns (and in fact intra-ns, too!) references the same way other authz is done...

This is a sig-auth KEP, so they need to own that analysis.   I don't want to bite off more than we can chew, but...

Author: Moved to "open questions" since I think we're going to need more discussion here.

	// Namespace is the namespace of the referent.

> 	Namespace string `json:"namespace"`

Reviewer 5: Are we sure we do not want a label selector here?  `kubernetes.io/metadata.name` can be used to target a single namespace.  Mutating namespaces in any way is considered a privileged action, so why is this limited to a hard coded name?

Reviewer 6: I'm on the fence with namespace selectors... that makes evaluation of the applicability of a reference grant significantly more difficult (have to read/watch changes to namespace labels, etc). I'd be curious to hear the experience of using namespace selectors in network policy.

hoisting from [sig-auth discussion](https://docs.google.com/document/d/1woLGRoONE3EBVx-wTb4pvp4CI7tmLZ6lS26VTbosLKM/edit#bookmark=id.9q15hi5lip46):

is "allow from all namespaces" possible/desirable to model? I tend to agree namespace selectors make it possible to overgrant accidentally, but for places that intend to allow cluster-wide reference (e.g. reference secret or config map containing trust roots from any route), is something like `namespace: "*"` plausible

Reviewer 3: network policy is built (and AdminNetPolicy continues to build) for a notion of "tenant" - defined as a set of NS labels.

Reviewer 3: > is "allow from all namespaces" possible/desirable to model?

At some point there was a desire to allow TLS certs and other things that are "cluster global", so yes.

Author: Enabling namespace boundaries to be crossed already introduces some risk, but going a step further and making it easy to allow references from _all namespaces_ feels especially risky. I would rather require namespaces to be individually specified, but I may be too narrowly looking at this from the perspective of Gateway API use cases.

Reviewer 2: I think that allowing access _from_ all namespaces is okay as long as you have to _explicitly opt in_.

It should require either a special value (Contour's TLSCertificateDelegation used `*` for `namespace` for this reason), or an explicit "allow from all namespaces" field, which could have BIG TEXT CALLOUTS about being careful.

Reviewer 2: We also skipped a label selector here because of the watch complexity it adds for implementations - keeping this as a string per `From` entry means that only updates to the ReferenceGrant itself can cause changes in the auth outcome, whereas doing a selector means that both ReferenceGrant and Namespace label changes need to be watched.

I accept that this is not a good reason for keeping this around forever though. 😄

Reviewer 3: The problem I have with this is that there's almost nothing else which works this way - most everything else is selectors.

You might go look at https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/2091-admin-network-policy, specifically `AdminNetworkPolicyPeer`.  I'm not saying we need to adopt all of this off the bat, but it does seem like some notion of "same tenant" is going to be common.

e.g. allow references to all Services from any namespace which has the same value of label "example.com/team" as I do.  Maybe the indirection is not needed ("same as me") but "any namespace where example.com/team = myteam" seems totally reasonable?

Author: >The problem I have with this is that there's almost nothing else which works this way - most everything else is selectors.

If we consider this API as being more similar to RBAC than NetworkPolicy, a `*` would at least be consistent with the existing authorization APIs Kubernetes supports.

If we adopted the additional `NamespaceResource` struct would we also need to have additional nesting in the YAML to enable the future expansion in a backwards compatible way? 

Although we could make the case that a selector would be useful in place of both the namespace and resource names, I'm concerned about how this would increase complexity for both implementations and users. Leaving a path to allow that in the future is probably smart, but it would be unfortunate to leave an extra level of nesting in the API if it never ended up being needed.

Reviewer 3: > would we also need to have additional nesting in the YAML to enable the future expansion in a backwards compatible way?

Yes.  In general you want a one-of to be a distinct struct, which means extra indent.

> it would be unfortunate to leave an extra level of nesting in the API if it never ended up being needed

tradeoffs, tradeoffs.. :)

Author: Moved to "open questions" since I think we're going to need more discussion here.

}

// ReferenceGrantTo describes what Kinds are allowed as targets of the
// references.
type ReferenceGrantTo struct {
	// Group is the group of the referent.
	// When empty, the Kubernetes core API group is inferred.
	Group string `json:"group"`

	// Kind is the kind of the referent.
	Kind string `json:"kind"`

	// Name is the name of the referent. When unspecified, this policy
	// refers to all resources of the specified Group and Kind in the local
	// namespace.
	//
	// +optional

> 	Name *string `json:"name,omitempty"`

Reviewer 3: why name and not a selector?  Or a list of names?

Author: You can omit name to allow references to every resource of the specified Kind in a namespace. Cross-Namespace references are easy to mess up, I'd hate to further complicate this with label selectors. Not only would this complicate the implementation, I think it would result in confusing configuration. As far as a list of names, `To` is already a list, I'd rather avoid adding a list within a list.

Reviewer 2: Yes, the design intent here is that separate rules are separate entries in the `To` list.

Reviewer 9: In the case of VolumeSnapshots, a common usage pattern is to periodically take and rotate snapshots (every few hours, once a day, etc), where explicitly enumerating each would be cumbersome. With this model, to allow a single or subset of snapshots, some other automation would be required to rotate the names as the snapshots get rotated.

Reviewer 9: Any should work for most use cases. My scenario of a single/subset was more hypothetical where someone may not want to expose all snapshots. And even so, they can express it with the proposal, it would just be some additional work to rotate the names.

Reviewer 3: I suspect "any" would work, too, though we can certainly fabricate cases where a selector is needed.

}
```

### Outstanding questions and clarifications

This section lists some of the outstanding questions people have had about
ReferenceGrant, and other items that we'll be clarifying in the godoc and other
documentation as part of the transition to the new API group, keeping in mind
that we will not be adding any new features, fields, or behavior, just clarifying

> what's done already.

Reviewer 2: Yes, with the feedback we've already got, keeping the resource the same is not viable. I'll edit this accordingly.

Also note that we don't consider any of these blockers for the general _idea_ of
moving ReferenceGrant to the new API group, just notes to save discussion time.

* Clarify that an implementation is required to reconcile ReferenceGrant for
  specific `To` Kinds.
* Corollary for future work, define how controllers interact. Is it a problem if
  multiple controllers consume the same ReferenceGrant?

> * Status design is still pending, but it's currently expected that controllers

Reviewer 3: Agree - I am not convinced that spec/status is meaningful here

Author: Resolving this because there are other threads that cover status discussion more thoroughly.

  will indicate status on the _referring_ resources, not on ReferenceGrant itself.
* Clarify that the expected operating model for implementations expects them to
  have broad, read access to both ReferenceGrant and the specific `To` Kinds they
  support, and then self-limit to only _use_ the relevant ones.
* Decide whether to formally add `*` as a special value for Namespace, to mean
  "all namespaces".

### Test Plan

[x] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

This is a net new resource to Kubernetes so it will not require any changes or
additions to existing tests.

##### Unit tests

Unit tests will be used to cover:

1. ReferenceGrant validation

> 2. ReferenceGrant implementation library

Author: Thanks! Used these as a baseline for test plan.

More details will be added as the details of the implementation library are

> clarified.

Reviewer 6: hoisting testcases from the [sig-auth discussion](https://docs.google.com/document/d/1woLGRoONE3EBVx-wTb4pvp4CI7tmLZ6lS26VTbosLKM/edit#bookmark=id.9q15hi5lip46):

more an implementation question, but is the result of the following scenarios observably different? (spoiler: I don't think it should be observably different or we leak information about existence of namespaces/objects):
* reference to namespace that doesn't exist
* reference to namespace that exists, instance that doesn't exist
* reference to namespace that exists, instance that exists, covering ReferenceGrant doesn't exist (looks like a RefNotPermitted condition is added?)

Author: Good points, these are all intended to be equivalent. Added clarification both in unit test details and in proposed API Spec.

##### Integration tests

Because we are not bundling an implementation with this API, no integration

> tests are possible.

Reviewer 6: it depends on the details of the library, but if responding to revocations of reference grants involves watches, integration tests of the library could make sense

Author: Added proposal for integration tests, going to resolve this and we can follow up on the text of that proposal

> tests are possible.

Reviewer 5: For example, can we provide a complete example/reference implementation as we have done for aggregated API servers, client go credential plugins, etc and have tests that show it all works together?

> tests are possible.

Reviewer 8: Perhaps we can define an example CRD.

Eg: the PairedExample CR relies on a ReferenceGrant to pair with an PairedExample in another namespace. We also provider a sample controller that works with it to remove the one side of the pair if either the other disappears or the grant is lost?

IMO we wouldn't need that test until we're targeting GA.

> tests are possible.

Author: Good idea, added details to the KEP.

##### e2e tests

We will strongly encourage every API that uses ReferenceGrant to define
conformance tests for their use of ReferenceGrant.

### Graduation Criteria

#### GA

[x] Almost all of the fields and behavior have conformance test coverage.

> [x] Multiple conformant implementations.

Reviewer 6: this doesn't seem like it can be checked quite yet :)

Author: Woops, thanks, it is true of the current conformance test coverage of the Gateway API resource, but given the changes this KEP will result in, it definitely is not true now, will update.

[x] Widespread implementation and usage.
[ ] Conformance tests that exercise all ReferenceGrant API calls (not the actual implementation of the API).
[ ] At least 6 months of soak time as a beta API (graduated to beta in Dec 2022).

### Upgrade / Downgrade Strategy

N/A

### Version Skew Strategy

> Version skew is a bit different here since it won't be implemented in-tree.

Reviewer 6: this bit probably needs updating now that this is proposing being in-tree, specifically around the library implementation and how it supports different versions / falls back if preferred versions are unavailable

Author: Tried to clarify this to make the distinction that the actual implementing controllers of this API will not be in-tree, even if they are importing a library from in-tree. As far as fall backs, do we have any standard guidelines here? This feels like it would be very similar to any other API.

restarted to pick up the presence of this API.

###### Are there any tests for feature enablement/disablement?

No

### Rollout, Upgrade and Rollback Planning

###### How can a rollout or rollback fail? Can it impact already running workloads?

API enablement may not work, but that would not be unique to this API.

###### What specific metrics should inform a rollback?

N/A, this is just an API

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

N/A, this is just an API

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

No

### Monitoring Requirements

N/A, this is just an API

###### How can an operator determine if the feature is in use by workloads?
```
kubectl get referencegrants --all-namespaces
```

###### How can someone using this feature know that it is working for their instance?

This will be dependent on the API that ReferenceGrant is used with. In Gateway API,
each resource has clear status conditions that reflect the validity of a cross-namespace
reference.

###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?

> N/A, this is just an API

Reviewer 6: if there's a library provided, and expectation of reacting to revocation of reference grants, a possible SLO would be time between revocation and notification by the library

Author: Good call, added.

###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?

N/A, this is just an API

###### Are there any missing metrics that would be useful to have to improve observability of this feature?

No

### Dependencies

N/A, this is just an API

###### Does this feature depend on any specific services running in the cluster?

- API Server

### Scalability

###### Will enabling / using this feature result in any new API calls?

Yes, users may install controllers that watch for changes to ReferenceGrants.
Users may also create ReferenceGrants to enable cross-namespace references.

###### Will enabling / using this feature result in introducing new API types?

API Type: ReferenceGrant
Supported Number of Objects per Cluster: No limit 

> Supported Number of Objects per Namespace: No limit

Reviewer 6: not sure if there are global custom resource limits that sig-api-machinery or sig-scalability would prefer be added here

removal of the ReferenceGrant will also result in the removal of the HTTPRoute's
config from the data plane.
#### PersistentVolumeClaim using cross namespace data source

This example is taken from https://kubernetes.io/blog/2023/01/02/cross-namespace-data-sources-alpha/
and updated to use the proposed new spec.

It allows the PersistentVolumeClaim in the `dev` namespace to use a volume
snapshot from the `prod` namespace as its data source.

```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: example-pvc
  namespace: dev
spec:
  storageClassName: example
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 1Gi
  dataSourceRef:
    apiGroup: snapshot.storage.k8s.io
    kind: VolumeSnapshot
    name: new-snapshot-demo
    namespace: prod
  volumeMode: Filesystem
---
apiVersion: authorization.k8s.io/v1alpha1
kind: ReferenceGrant
metadata:
  name: allow-prod-pvc
  namespace: prod
spec:
  from:
  - resource: persistentvolumeclaims
    namespace: dev
  to:
  - group: snapshot.storage.k8s.io
    resource: volumesnapshots
    name: new-snapshot-demo
```

At time of writing, I'm not sure about the behavior here when the ReferenceGrant

> is deleted.

Reviewer 2: @msau42, I didn't know where to go to find this out, if you could comment here, I'll update with any details you can provide.

Author: @ttakahashi21 can you help provide more context here?

Reviewer 10: https://github.com/kubernetes/enhancements/pull/3767/files#r1084757612
It works as described in the comment above.
It is only used when creating a PVC. So, revoking access would not impact the existing volume.
Only new volumes would be rejected.

Reviewer 9: I think the comment got lost after uploading a new set of changes. Copying it over here:

For the volume provisioning from datasource scenario, once the volume is provisioned, it no longer actively uses the reference, so revoking access would not impact the existing volume. Only new volumes would be rejected.

Author: Thanks, clarified in the KEP, PTAL

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

[ReferenceGrant](https://gateway-api.sigs.k8s.io/api-types/referencegrant/) was
developed by the [Gateway API subproject](https://gateway-api.sigs.k8s.io/) to
enable certain object references to cross namespaces. More recently, it has also
been [used by
sig-storage](https://kubernetes.io/blog/2023/01/02/cross-namespace-data-sources-alpha/)
to enable cross-namespace data sources.

This KEP proposes moving ReferenceGrant from its current
`gateway.networking.k8s.io` API group into the `authorization.k8s.io` API
group.

## Motivation

Any project that wants to enable cross-namespace references currently has to choose
between introducing a dependency on Gateway API's ReferenceGrant or creating a
new API that would be partially redundant (leading to confusion for users).
	
Recent interest between SIGs has made it clear that ReferenceGrant is wanted for use
cases other than Gateway API. We would like to move ReferenceGrant to a neutral home

> (ideally, under sig-auth) in order to make it the canonical API for managing references

Reviewer 8: (nit)
```suggestion
(ideally, under SIG Auth) in order to make it the canonical API for managing references
```

Author: Updated, thanks!

across namespaces.

### Goals

> * Move ReferenceGrant to sig-auth API Group

Reviewer 8: ```suggestion
* Move ReferenceGrant to an API Group that SIG Auth manages
```

Author: Updated, thanks!

* Clearly define how ReferenceGrant should be used, including both current use
  cases and guidance for future use cases
* Implement a library to ensure that ReferenceGrant is implemented consistently
  by all controllers

### Non-Goals

* Add, remove, or change fields in the ReferenceGrant API
* Develop an authorizer that will automatically implement ReferenceGrant for
  all use cases

## Proposal

Move the existing ReferenceGrant resource into a new
`authorization.k8s.io` API group, defined within the Kubernetes code base
as part of the 1.27 release.

We will take this opportunity to clarify and update the API after SIG-Auth

> feedback. This resource will start with v1alpha1 as the API version.

Reviewer 8: Will we also provide an official CRD for people who want to use the new API on older clusters?

Reviewer 2: I like the sound of this, but I don't know anything about how it would work. 😄

Author: I think the idea that this resource will also include an authorization layer that insures that the creator of the resource also has read or write access to the resource they're granting access to would make this pretty tough. Instead I think we'd just recommend using the CRD already provided by Gateway API, which unfortunately will be slightly different than what SIG Auth accepts here.

### General Notes

#### `ReferenceGrant` is half of a handshake

When thinking about ReferenceGrant, it is important to remember that it does not
do anything by itself. It *Grants* the *possibility* of making a *Reference*
across namespaces. It's intended that _another object_ (that is, the From object)
complete the handshake by creating a reference to the referent object (the To
object).

#### `Resource` vs `Kind`

When creating a metaresource (that is, a resource that targets other resources)
like ReferenceGrant, it's important to consider if the metaresource uses the more
common `Kind` or the more correct `Resource`.

In the original Gateway API implementation, we chose to use `Kind` rather than
`Resource`, mainly to improve the user experience. That is, it's easier users
to take the value from the `kind` field at the top of the YAML they are already
using, and put it straight into these fields, rather than needing to do a
kind-resource lookup for every user's interaction with the API. @robscott even
ended up making https://github.com/kubernetes/community/pull/5973 to clarify
the API conventions.

However, in discussion on this KEP, it's clear that the more generic nature of

> _this_ API requires the additional specificity that `Resource` provides.

Reviewer 8: > Add, remove, or change fields in the ReferenceGrant API

&mdash; non-goals, above

We can pick either; maybe we shouldn't have both though.

I'm fine with the API change, and I think we should revise the non-goal to clarify

Author: Good call, removed from non-goals. We really were trying to avoid API changes, but it looks like we're going to be stuck with at least the kind -> resource one.

The Gateway API ReferenceGrant looked like this:
```yaml
apiVersion: gateway.networking.k8s.io/v1beta1
kind: ReferenceGrant
metadata:
  name: allow-gateways
  namespace: bar
spec:
  from:
    # Note that in Gateway API, Group is currently defaulted
    # to this, which means you to explicitly set the group to
    # the empty string for Core resources. We should definitely
    # change this.
    - group: "gateway.networking.kubernetes.io"
      kind: Gateway
      namespace: foo
  to:
   - group: ""
     kind: Secret
```

The new version will look like this instead:
```yaml
apiVersion: authorization.k8s.io/v1alpha1
kind: ReferenceGrant
metadata:
  name: allow-gateways
  namespace: bar
spec:
  from:
    # Assuming that we leave the default for Group to the empty
    # string, so that Core objects don't need additional config.
    - group: "gateway.networking.kubernetes.io" 
      resource: gateways
      namespace: foo
  to:
    - resource: secrets

```

> The new version communicates the scope more clearly because `resource` is plural.

Author: Good point, thanks, updated!

> The new version communicates the scope more clearly because `resource` is plural.

Reviewer 2: I meant that the pluralness makes things a little easier for the user who may not be one hundred percent clear on the details of the resource/kind distinction. I'm also not a fan, but here we are.

#### Revocation behavior

Unfortunately, there's no way to be specific about what happens when a
ReferenceGrant is deleted in every possible case - the revocation behavior is

> dependent on what access is being granted (and revoked).

Reviewer 8: However, we _can_ document what we behaviors we expect implementer to follow.

Author: Good call, updated

}

// ReferenceGrantSpec identifies a cross namespace relationship that is trusted
// for Gateway API.
type ReferenceGrantSpec struct {
	// From describes the trusted namespaces and kinds that can reference the
	// resources described in "To". Each entry in this list MUST be considered
	// to be an additional place that references can be valid from, or to put
	// this another way, entries MUST be combined using OR.
	//
	// +kubebuilder:validation:MinItems=1
	// +kubebuilder:validation:MaxItems=16
	From []ReferenceGrantFrom `json:"from"`

	// To describes the resources in this namespace that may be referenced by
	// the resources described in "From". Each entry in this list MUST be
	// considered to be an additional set of objects that references can be
	// valid to, or to put this another way, entries MUST be combined using OR.
	//
	// +kubebuilder:validation:MinItems=1
	// +kubebuilder:validation:MaxItems=16
	To []ReferenceGrantTo `json:"to"`
}

// ReferenceGrantFrom describes trusted namespaces and kinds.
type ReferenceGrantFrom struct {
	// Group is the group of the referent.
	// When empty, the Kubernetes core API group is inferred.

> 	Group Group `json:"group"`

Author: We're already opening up a lot by enabling cross-namespace references, I want this to be as tightly defined as possible so this API doesn't become unnecessarily dangerous. I think in that case the idea was that someone may want to allow references from any resource in the Gateway API group. Of course that API group keeps expanding and granting access that broadly could result in unintended exposure if/when those new kinds of resources are added to a cluster.

> 	Group Group `json:"group"`

Reviewer 3: This is a hard to revisit decision once it goes beta, so ... tick tock tick tock...  I'd leave this to a sig-auth approver.

> 	Group Group `json:"group"`

Author: Moved to "open questions" since I think we're going to need more discussion here.

> 	Group Group `json:"group"`

Reviewer 6: I didn't see this question listed in the open questions... did I miss it?

> 	Group Group `json:"group"`

Reviewer 6: (my instinct is to require group and resource, fwiw)

> 	Group Group `json:"group"`

Author: Added link to this discussion the "Do We Need Any or Wildcard Selectors?" open question. That specifically asks:

Should we expand that to allow any of the following?

1. References to any group or resource
1. References from any group or resource
1. References from any namespace

	// Kind is the resource of the referent.
	Resource string `json:"resource"`

	// Namespace is the namespace of the referent.
	Namespace string `json:"namespace"`
}

// ReferenceGrantTo describes what Kinds are allowed as targets of the
// references.
type ReferenceGrantTo struct {
	// Group is the group of the referent.
	// When empty, the Kubernetes core API group is inferred.

> 	Group string `json:"group"`

Author: I think this is mostly the same discussion as "From". In both cases I'd strongly prefer requiring these to be specified. We're already opening up a lot by enabling cross-namespace references, I want this to be as tightly defined as possible so this API doesn't become unnecessarily dangerous. I think in that case the idea was that someone may want to allow references from any resource in the Gateway API group. Of course that API group keeps expanding and granting access that broadly could result in unintended exposure if/when those new kinds of resources are added to a cluster.

> 	Group string `json:"group"`

Author: Thinking about it more, I think "any" becomes more acceptable if we're clearly defining the verb(s) we're allowing. My main concerns were about unintentionally allowing an undefined and unknown set of actions that could grow in the future. If we're restricting the kinds of actions that we're allowing, this is no longer relevant.

> 	Group string `json:"group"`

Author: Moved to "open questions" since I think we're going to need more discussion here.
